### PR TITLE
Add group and label to hover tooltip if provided by user

### LINKED ISSUE #3603: Include label in hover tool
It would be nice to be able to easily include the label for a given, e.g., `Curve` in the hovertool. I tried doing so like this

```python
import numpy as np
import holoviews as hv
from bokeh.models import HoverTool

hv.extension("bokeh")
hover = HoverTool(tooltips=[("label", "$name"), ("x", "@x"), ("y", "@y")])
hv.opts.defaults(hv.opts.Curve(tools=[hover]))

x = np.linspace(-5, 5, 101)
hv.Curve((x, np.sin(x)), label="sin") * hv.Curve((x, np.cos(x)), label="cos")
```

but my hovertool looks like:

image

It does seem like there is a `name` attribute, because if I do

```python
hover = HoverTool(tooltips=[("label", "$random_thing"), ("x", "@x"), ("y", "@y")])
```

instead, the hovertool doesn't work at all. Perhaps the label from the `hv.Curve` needs to be passed back to a bokeh glyph for this to work? Or is there a different way that I could reference the label in the hovertool?

### PR DESCRIPTION
**Pull request by droumis**
Adds the group and label args to the bokeh hover tooltip if provided by user.

It seems to me like a good idea to have this enabled by default, but I'm guessing that there's a reason why it wasn't already in place. Opinions welcome.

Fixes https://github.com/holoviz/holoviews/issues/3603

<details> <summary> Code </summary>

```python
import numpy as np
import panel as pn
import holoviews as hv
hv.extension('bokeh')

x = np.linspace(0, 10*np.pi)
curves = {}
j = 0.5
yvals = j * np.sin(x)

curves = []
curves.append(hv.Curve((x, yvals)).opts(tools=['hover'], color='blue'))
curves.append(hv.Curve((x + 2 * np.pi/2, 2 * yvals), label='curve2').opts(tools=['hover'], color='red'))
curves.append(hv.Curve((x + 3 * np.pi/2, 3 * yvals), group='A').opts(tools=['hover'], color='green'))
curves.append(hv.Curve((x + 4 * np.pi/2, 4 * yvals), label='curve4', group='A').opts(tools=['hover'], color='orange'))
pn.Row(hv.Overlay(curves).opts(width=600, height=600)).servable()

```

</details>

https://github.com/holoviz/holoviews/assets/6613202/659e42fb-d82e-47f4-8895-18ad357e6282

TODO:
- [x] add tests

### CODE REVIEW

**Review by hoxbro — COMMENTED** [review]
> It seems to me like a good idea to have this enabled by default, but I'm guessing that there's a reason why it wasn't already in place. 

A reason why it hasn't been implemented can be as simple as there has been no demand for it. 

> I don't really have high hopes for this getting merged

You shouldn't put down the effort you have put into the PR. If you don't believe in it yourself, it is hard to get others to believe in you :+1: 

---

I would like to see a unit test. To get the tooltips after creation you can play look around with ` hv.renderer('bokeh').get_plot(hv.Curve([])).handles` (in test we have `bokeh_renderer` as shorthand for the first part). 

---

This has nothing to do with this PR, but the legend seems wrong for your example.

**Comment by droumis** [comment]
> You shouldn't put down the effort you have put into the PR. If you don't believe in it yourself, it is hard to get others to believe in you 👍

Thanks, I adjusted my language.

File: holoviews/plotting/bokeh/element.py [hunk 1]
```diff
@@ -295,6 +305,7 @@ def _init_tools(self, element, callbacks=None):
         tooltips, hover_opts = self._hover_opts(element)
         tooltips = [(ttp.pprint_label, '@{%s}' % util.dimension_sanitizer(ttp.name))
                     if isinstance(ttp, Dimension) else ttp for ttp in tooltips]
+        tooltips += self._group_label(element)
```
[the remark this answers is not in the document]
Comment (droumis) [reply holoviews/plotting/bokeh/element.py:308]: not sure I understand this part. Aren't I already doing the check in the method whether the data contains `label` or `group`? Are you suggesting that the method should return an unaltered tooltips list rather than an empty list, if `label` or `group` are not defined by the user?

**Comment by droumis** [comment]
> This has nothing to do with this PR, but the legend seems wrong for your example.

What's wrong about it? seems accurate to me and reflects the labels of the curves for which a label was given

File: holoviews/plotting/bokeh/element.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hoxbro) [reply holoviews/plotting/bokeh/element.py:308]: If the data has a `label` or `group` dimension, that should take precedence over the elements. An example is the following:

``` python
import holoviews as hv
import pandas as pd
hv.extension("bokeh")

hv.Curve(pd.DataFrame(dict(label=range(4)))).opts(tools=['hover'])
```

**Comment by Hoxbro** [comment]
> What's wrong about it? seems accurate to me and reflects the labels of the curves for which a label was given

It looks like it has a blue border at the top. But I can't recreate it locally, so just ignore that comment.  
Screenshot 2024-02-21 16 55 49

File: holoviews/plotting/bokeh/element.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (droumis) [reply holoviews/plotting/bokeh/element.py]: ah ok I understand now, thanks for clarifying

[the remark this answers is not in the document]
Comment (droumis) [reply holoviews/plotting/bokeh/element.py]: since the tooltip dims are already sanitized, I think this can simply be done by passing in tooltips as you say and then adding this to each conditional:

`'label' not in [t[0] for t in tooltips]`

[the remark this answers is not in the document]
Comment (droumis) [reply holoviews/plotting/bokeh/element.py]: <details> <summary> Code </summary>

```python
import numpy as np
import panel as pn
import pandas as pd
import holoviews as hv
hv.extension('bokeh')

x = np.linspace(0, 10*np.pi)
curves = {}
j = 0.5
yvals = j * np.sin(x)
curves = []
curves.append(hv.Curve((x, yvals)).opts(tools=['hover'], color='blue'))
curves.append(hv.Curve((x + 2 * np.pi/2, 2 * yvals), label='curve2').opts(tools=['hover'], color='red'))
curves.append(hv.Curve((x + 3 * np.pi/2, 3 * yvals), group='A').opts(tools=['hover'], color='green'))
curves.append(hv.Curve((x + 4 * np.pi/2, 4 * yvals), label='curve4', group='A').opts(tools=['hover'], color='orange'))
curves.append(hv.Curve(pd.DataFrame(dict(label=range(4))), label='no_label_ttip').opts(tools=['hover'], color='purple'))
pn.Row(hv.Overlay(curves).opts(width=600, height=600)).servable()
```

</details>


https://github.com/holoviz/holoviews/assets/6613202/87001830-3da0-4641-9c4b-e47cad173534

**Comment by droumis** [comment]
Tests added

File: holoviews/tests/plotting/bokeh/test_group_label_tooltips.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import numpy as np
```
Comment (Hoxbro) [inline holoviews/tests/plotting/bokeh/test_group_label_tooltips.py:1]: I think we can move this test into bokeh/test_elementplot.py

File: holoviews/plotting/bokeh/element.py [hunk 3]
```diff
@@ -286,6 +286,17 @@ def _hover_opts(self, element):
...
+    def _add_element_group_and_label_hover_tooltips(self, element, tooltips):
+        # return group and label if present
+        group_label = []
+        # group is the element type by default (e.g. Curve)
+        # skip if 'group' and 'label' are data dims that are already in tooltips
+        if hasattr(element, 'group') and element.group != element.param.group.default and 'group' not in [t[0].lower() for t in tooltips]:
+            group_label.append(('Group', element.group))
+        if hasattr(element, 'label') and element.label and 'label' not in [t[0].lower() for t in tooltips]:
+            group_label.append(('Label', element.label))
+        return group_label + tooltips
```
Comment (Hoxbro) [inline holoviews/plotting/bokeh/element.py:298]:
```suggestion
    def _add_element_group_and_label_hover_tooltips(self, element, tooltips):
       """ 
       Add elements group and label to the hover tooltip if they have changed from default.
       Will not add elements group or label if it is part of data dimension
       """
        if hasattr(element, 'group') and element.group != element.param.group.default and not any('group' == t[0].lower() for t in tooltips):
            tooltips.append(('Group', element.group))
        if hasattr(element, 'label') and element.label and not any('label' == t[0].lower() for t in tooltips):
            tooltips.append(('Label', element.label))
```

This change does 3 things:
1) Make comment docstring. This will make it show up with an LSP.
2) Add to the existing tooltip instead of creating a new tooltip (this will also need a change in the next function) as we don't return anything anymore
3) Exit early if 'group' or 'label' are present in the tooltips.

What do you think about the changes?

(This is written as a suggestion, so please double-check the suggestion)

File: holoviews/tests/plotting/bokeh/test_group_label_tooltips.py (hunk 2 above)
Comment (droumis) [reply holoviews/tests/plotting/bokeh/test_group_label_tooltips.py]: I added a new section in the existing class for hover tooltip tests. hope that's what you meant

File: holoviews/plotting/bokeh/element.py (hunk 3 above)
Comment (droumis) [reply holoviews/plotting/bokeh/element.py]: Thanks for the guidance re: not any(...) to end early, and generally about mutating in place if possible. That's all very helpful.

Me thinks Group and Label should be displayed first in the hover tooltips, which is why I was prepending rather than appending. The hierarchy being: 1. Groups of Elements 2. Label of an entire element 2. Names of an element's individual data dims. So how about the following to achieve in-place prepending, I just pushed this:

```python
def _add_element_group_and_label_hover_tooltips(self, element, tooltips):
       """ 
       Prepend the element's group and label to the hover tooltip if they have changed from default.
       Will not prepend element's group or label if 'group' or 'label' exist as data dimensions
       """
        if hasattr(element, 'label') and element.label and not any('label' == t[0].lower() for t in tooltips):
            tooltips.insert(0, ('Label', element.label))
        if hasattr(element, 'group') and element.group != element.param.group.default and not any('group' == t[0].lower() for t in tooltips):
            tooltips.insert(0, ('Group', element.group))
```

**Comment by Hoxbro** [comment]
One last thing: Can you try setting the group/label to something like `hel lo`, `{hello` (testing for edge-cases)

**Comment by droumis** [comment]
> One last thing: Can you try setting the group/label to something like `hel lo`, `{hello` (testing for edge-cases)

added a test for whitespace and special chars 

image

**Comment by maximlt** [comment]
Just tried hvPlot from that branch.

Before:
image

After:
image

I think it's coming from these lines:
https://github.com/holoviz/hvplot/blob/f17f4705afe40f4c8e4e388823004b14ec938519/hvplot/converter.py#L1567-L1568

There are maybe more cases, I just did a very quick check.

**Comment by ahuang11** [comment]
Maybe if isinstance element, 2D, then don't include the label?

**Comment by droumis** [comment]
At a HoloViz meeting, there was discussion about supporting a better syntax for a custom hover tool.

Here is the unearthed thread on the topic: https://github.com/holoviz/holoviews/issues/1816

I would prefer a more fully featured custom hover tool syntax as originally proposed, as long as it supports adding arbitrary entries like 'label' and 'group' arg values, in addition to standard data dims. 

I will close this current PR, unmerged.
